File: reaktor-ui/ts/src/tokens/DesignTokens.ts
/**
 * Reaktor UI Design Token System for React/TypeScript
 *
 * This mirrors the Kotlin token system for cross-platform consistency.
 * All values should match the Compose Multiplatform implementation.
 */

// ============================================================================
// COLOR TYPES
// ============================================================================

export interface ColorSchemeTokens {
  // Surface colors
  background: string;
  surface: string;
  surfaceVariant: string;
  surfaceContainer: string;
  surfaceContainerHigh: string;
  surfaceContainerLow: string;

  // Content colors
  onBackground: string;
  onSurface: string;
  onSurfaceVariant: string;

  // Primary
  primary: string;
  primaryContainer: string;
  onPrimary: string;
  onPrimaryContainer: string;

  // Secondary
  secondary: string;
  secondaryContainer: string;
  onSecondary: string;
  onSecondaryContainer: string;

  // Tertiary
  tertiary: string;
  tertiaryContainer: string;
  onTertiary: string;
  onTertiaryContainer: string;

  // Feedback
  error: string;
  errorContainer: string;
  onError: string;
  onErrorContainer: string;

  success: string;
  onSuccess: string;

  warning: string;
  onWarning: string;

  info: string;
  onInfo: string;

  // Utility
  outline: string;
  outlineVariant: string;
  scrim: string;
  shadow: string;

  // Inverse
  inverseSurface: string;
  inverseOnSurface: string;
  inversePrimary: string;
}

// ============================================================================
// TYPOGRAPHY TYPES
// ============================================================================

export type FontWeight =
  | 'thin' | 'extraLight' | 'light' | 'normal'
  | 'medium' | 'semiBold' | 'bold' | 'extraBold' | 'black';

export interface TextStyleTokens {
  fontSize: string;
  lineHeight: string;
  fontWeight: FontWeight;
  letterSpacing: string;
}

export interface TypographyTokens {
  displayLarge: TextStyleTokens;
  displayMedium: TextStyleTokens;
  displaySmall: TextStyleTokens;

  headlineLarge: TextStyleTokens;
  headlineMedium: TextStyleTokens;
  headlineSmall: TextStyleTokens;

  titleLarge: TextStyleTokens;
  titleMedium: TextStyleTokens;
  titleSmall: TextStyleTokens;

  bodyLarge: TextStyleTokens;
  bodyMedium: TextStyleTokens;
  bodySmall: TextStyleTokens;

  labelLarge: TextStyleTokens;
  labelMedium: TextStyleTokens;
  labelSmall: TextStyleTokens;
}

// ============================================================================
// SPACING TYPES
// ============================================================================

export interface SpacingTokens {
  none: string;
  xxs: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  xxl: string;
  xxxl: string;
}

// ============================================================================
// SHAPE TYPES
// ============================================================================

export interface ShapeTokens {
  none: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  full: string;
}

// ============================================================================
// ELEVATION TYPES
// ============================================================================

export interface ElevationTokens {
  none: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
}

// ============================================================================
// SIZING TYPES
// ============================================================================

export interface SizingTokens {
  touchTargetMin: string;
  iconXs: string;
  iconSm: string;
  iconMd: string;
  iconLg: string;
  iconXl: string;
  buttonSm: string;
  buttonMd: string;
  buttonLg: string;
  inputSm: string;
  inputMd: string;
  inputLg: string;
  avatarSm: string;
  avatarMd: string;
  avatarLg: string;
  avatarXl: string;
}

// ============================================================================
// BREAKPOINT TYPES
// ============================================================================

export interface BreakpointTokens {
  mobile: number;
  tablet: number;
  desktop: number;
  largeDesktop: number;
}

// ============================================================================
// MOTION TYPES
// ============================================================================

export interface MotionTokens {
  durationInstant: number;
  durationFast: number;
  durationNormal: number;
  durationSlow: number;
  durationSlowest: number;
}

// ============================================================================
// COMBINED DESIGN TOKENS
// ============================================================================

export interface DesignTokens {
  colors: ColorSchemeTokens;
  typography: TypographyTokens;
  spacing: SpacingTokens;
  shapes: ShapeTokens;
  elevation: ElevationTokens;
  sizing: SizingTokens;
  breakpoints: BreakpointTokens;
  motion: MotionTokens;
}

// ============================================================================
// DEFAULT VALUES
// ============================================================================

export const defaultTypography: TypographyTokens = {
  displayLarge: { fontSize: '57px', lineHeight: '64px', fontWeight: 'normal', letterSpacing: '0' },
  displayMedium: { fontSize: '45px', lineHeight: '52px', fontWeight: 'normal', letterSpacing: '0' },
  displaySmall: { fontSize: '36px', lineHeight: '44px', fontWeight: 'normal', letterSpacing: '0' },

  headlineLarge: { fontSize: '32px', lineHeight: '40px', fontWeight: 'normal', letterSpacing: '0' },
  headlineMedium: { fontSize: '28px', lineHeight: '36px', fontWeight: 'normal', letterSpacing: '0' },
  headlineSmall: { fontSize: '24px', lineHeight: '32px', fontWeight: 'normal', letterSpacing: '0' },

  titleLarge: { fontSize: '22px', lineHeight: '28px', fontWeight: 'medium', letterSpacing: '0' },
  titleMedium: { fontSize: '16px', lineHeight: '24px', fontWeight: 'medium', letterSpacing: '0.15px' },
  titleSmall: { fontSize: '14px', lineHeight: '20px', fontWeight: 'medium', letterSpacing: '0.1px' },

  bodyLarge: { fontSize: '16px', lineHeight: '24px', fontWeight: 'normal', letterSpacing: '0.5px' },
  bodyMedium: { fontSize: '14px', lineHeight: '20px', fontWeight: 'normal', letterSpacing: '0.25px' },
  bodySmall: { fontSize: '12px', lineHeight: '16px', fontWeight: 'normal', letterSpacing: '0.4px' },

  labelLarge: { fontSize: '14px', lineHeight: '20px', fontWeight: 'medium', letterSpacing: '0.1px' },
  labelMedium: { fontSize: '12px', lineHeight: '16px', fontWeight: 'medium', letterSpacing: '0.5px' },
  labelSmall: { fontSize: '11px', lineHeight: '16px', fontWeight: 'medium', letterSpacing: '0.5px' },
};

export const defaultSpacing: SpacingTokens = {
  none: '0',
  xxs: '2px',
  xs: '4px',
  sm: '8px',
  md: '16px',
  lg: '24px',
  xl: '32px',
  xxl: '48px',
  xxxl: '64px',
};

export const defaultShapes: ShapeTokens = {
  none: '0',
  xs: '4px',
  sm: '8px',
  md: '12px',
  lg: '16px',
  xl: '24px',
  full: '9999px',
};

export const defaultElevation: ElevationTokens = {
  none: 'none',
  xs: '0 1px 2px rgba(0,0,0,0.05)',
  sm: '0 1px 3px rgba(0,0,0,0.1), 0 1px 2px rgba(0,0,0,0.06)',
  md: '0 4px 6px rgba(0,0,0,0.1), 0 2px 4px rgba(0,0,0,0.06)',
  lg: '0 10px 15px rgba(0,0,0,0.1), 0 4px 6px rgba(0,0,0,0.05)',
  xl: '0 20px 25px rgba(0,0,0,0.1), 0 10px 10px rgba(0,0,0,0.04)',
};

export const defaultSizing: SizingTokens = {
  touchTargetMin: '48px',
  iconXs: '16px',
  iconSm: '20px',
  iconMd: '24px',
  iconLg: '32px',
  iconXl: '48px',
  buttonSm: '32px',
  buttonMd: '40px',
  buttonLg: '48px',
  inputSm: '32px',
  inputMd: '40px',
  inputLg: '56px',
  avatarSm: '32px',
  avatarMd: '40px',
  avatarLg: '56px',
  avatarXl: '80px',
};

export const defaultBreakpoints: BreakpointTokens = {
  mobile: 0,
  tablet: 600,
  desktop: 840,
  largeDesktop: 1200,
};

export const defaultMotion: MotionTokens = {
  durationInstant: 0,
  durationFast: 150,
  durationNormal: 300,
  durationSlow: 500,
  durationSlowest: 700,
};

File: reaktor-ui/src/commonMain/kotlin/dev/shibasis/reaktor/ui/core/tokens/DesignTokens.kt
package dev.shibasis.reaktor.ui.core.tokens

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Reaktor UI Design Token System
 *
 * This is a platform-agnostic token system that can be implemented by any design system
 * (Material, Neomorphic, Glass, etc.). The base framework uses these tokens, not hardcoded values.
 *
 * Token hierarchy follows atomic design principles:
 * - Primitive tokens: Raw values (colors, sizes)
 * - Semantic tokens: Purpose-driven mappings (primary, surface, error)
 * - Component tokens: Component-specific values (button.padding, card.elevation)
 */

// ============================================================================
// PRIMITIVE COLOR TOKENS
// ============================================================================

/**
 * Raw color palette - no semantic meaning, just the palette available
 */
@Immutable
data class ColorPalette(
    val neutral0: Color = Color.White,
    val neutral10: Color = Color(0xFFF5F5F5),
    val neutral20: Color = Color(0xFFEEEEEE),
    val neutral30: Color = Color(0xFFE0E0E0),
    val neutral40: Color = Color(0xFFBDBDBD),
    val neutral50: Color = Color(0xFF9E9E9E),
    val neutral60: Color = Color(0xFF757575),
    val neutral70: Color = Color(0xFF616161),
    val neutral80: Color = Color(0xFF424242),
    val neutral90: Color = Color(0xFF212121),
    val neutral100: Color = Color.Black,

    val primary: Color = Color(0xFF6200EE),
    val primaryLight: Color = Color(0xFFBB86FC),
    val primaryDark: Color = Color(0xFF3700B3),

    val secondary: Color = Color(0xFF03DAC6),
    val secondaryLight: Color = Color(0xFF70EFDE),
    val secondaryDark: Color = Color(0xFF018786),

    val error: Color = Color(0xFFB00020),
    val errorLight: Color = Color(0xFFCF6679),
    val errorDark: Color = Color(0xFF8B0000),

    val success: Color = Color(0xFF4CAF50),
    val warning: Color = Color(0xFFFFC107),
    val info: Color = Color(0xFF2196F3),
)

// ============================================================================
// SEMANTIC COLOR TOKENS
// ============================================================================

/**
 * Semantic color scheme - colors with purpose, not just palette positions
 */
@Immutable
data class ColorSchemeTokens(
    // Surface colors
    val background: Color,
    val surface: Color,
    val surfaceVariant: Color,
    val surfaceContainer: Color,
    val surfaceContainerHigh: Color,
    val surfaceContainerLow: Color,

    // Content colors (text/icons on surfaces)
    val onBackground: Color,
    val onSurface: Color,
    val onSurfaceVariant: Color,

    // Primary action colors
    val primary: Color,
    val primaryContainer: Color,
    val onPrimary: Color,
    val onPrimaryContainer: Color,

    // Secondary action colors
    val secondary: Color,
    val secondaryContainer: Color,
    val onSecondary: Color,
    val onSecondaryContainer: Color,

    // Tertiary accent colors
    val tertiary: Color,
    val tertiaryContainer: Color,
    val onTertiary: Color,
    val onTertiaryContainer: Color,

    // Feedback colors
    val error: Color,
    val errorContainer: Color,
    val onError: Color,
    val onErrorContainer: Color,

    val success: Color,
    val onSuccess: Color,

    val warning: Color,
    val onWarning: Color,

    val info: Color,
    val onInfo: Color,

    // Utility colors
    val outline: Color,
    val outlineVariant: Color,
    val scrim: Color,
    val shadow: Color,

    // Inverse colors (for special containers)
    val inverseSurface: Color,
    val inverseOnSurface: Color,
    val inversePrimary: Color,
)

// ============================================================================
// TYPOGRAPHY TOKENS
// ============================================================================

/**
 * Typography scale - size and weight combinations
 */
@Immutable
data class TypographyTokens(
    // Display - largest, for hero text
    val displayLarge: TextStyleTokens = TextStyleTokens(57.sp, 64.sp, FontWeight.Normal),
    val displayMedium: TextStyleTokens = TextStyleTokens(45.sp, 52.sp, FontWeight.Normal),
    val displaySmall: TextStyleTokens = TextStyleTokens(36.sp, 44.sp, FontWeight.Normal),

    // Headline - section headers
    val headlineLarge: TextStyleTokens = TextStyleTokens(32.sp, 40.sp, FontWeight.Normal),
    val headlineMedium: TextStyleTokens = TextStyleTokens(28.sp, 36.sp, FontWeight.Normal),
    val headlineSmall: TextStyleTokens = TextStyleTokens(24.sp, 32.sp, FontWeight.Normal),

    // Title - smaller headers, card titles
    val titleLarge: TextStyleTokens = TextStyleTokens(22.sp, 28.sp, FontWeight.Medium),
    val titleMedium: TextStyleTokens = TextStyleTokens(16.sp, 24.sp, FontWeight.Medium),
    val titleSmall: TextStyleTokens = TextStyleTokens(14.sp, 20.sp, FontWeight.Medium),

    // Body - main content text
    val bodyLarge: TextStyleTokens = TextStyleTokens(16.sp, 24.sp, FontWeight.Normal),
    val bodyMedium: TextStyleTokens = TextStyleTokens(14.sp, 20.sp, FontWeight.Normal),
    val bodySmall: TextStyleTokens = TextStyleTokens(12.sp, 16.sp, FontWeight.Normal),

    // Label - buttons, captions
    val labelLarge: TextStyleTokens = TextStyleTokens(14.sp, 20.sp, FontWeight.Medium),
    val labelMedium: TextStyleTokens = TextStyleTokens(12.sp, 16.sp, FontWeight.Medium),
    val labelSmall: TextStyleTokens = TextStyleTokens(11.sp, 16.sp, FontWeight.Medium),
)

@Immutable
data class TextStyleTokens(
    val fontSize: TextUnit,
    val lineHeight: TextUnit,
    val fontWeight: FontWeight,
    val letterSpacing: TextUnit = 0.sp,
)

enum class FontWeight {
    Thin,
    ExtraLight,
    Light,
    Normal,
    Medium,
    SemiBold,
    Bold,
    ExtraBold,
    Black
}

// ============================================================================
// SPACING TOKENS
// ============================================================================

/**
 * Spacing scale - consistent spacing across the app
 */
@Immutable
data class SpacingTokens(
    val none: Dp = 0.dp,
    val xxs: Dp = 2.dp,
    val xs: Dp = 4.dp,
    val sm: Dp = 8.dp,
    val md: Dp = 16.dp,
    val lg: Dp = 24.dp,
    val xl: Dp = 32.dp,
    val xxl: Dp = 48.dp,
    val xxxl: Dp = 64.dp,
)

// ============================================================================
// SHAPE TOKENS
// ============================================================================

/**
 * Shape/corner radius scale
 */
@Immutable
data class ShapeTokens(
    val none: Dp = 0.dp,
    val xs: Dp = 4.dp,
    val sm: Dp = 8.dp,
    val md: Dp = 12.dp,
    val lg: Dp = 16.dp,
    val xl: Dp = 24.dp,
    val full: Dp = 9999.dp, // Fully rounded (pill shape)
)

// ============================================================================
// ELEVATION/SHADOW TOKENS
// ============================================================================

/**
 * Elevation levels for depth
 */
@Immutable
data class ElevationTokens(
    val none: Dp = 0.dp,
    val xs: Dp = 1.dp,
    val sm: Dp = 2.dp,
    val md: Dp = 4.dp,
    val lg: Dp = 8.dp,
    val xl: Dp = 16.dp,
)

// ============================================================================
// SIZING TOKENS
// ============================================================================

/**
 * Component size tokens for consistent sizing
 */
@Immutable
data class SizingTokens(
    // Touch target minimums
    val touchTargetMin: Dp = 48.dp,

    // Icon sizes
    val iconXs: Dp = 16.dp,
    val iconSm: Dp = 20.dp,
    val iconMd: Dp = 24.dp,
    val iconLg: Dp = 32.dp,
    val iconXl: Dp = 48.dp,

    // Button heights
    val buttonSm: Dp = 32.dp,
    val buttonMd: Dp = 40.dp,
    val buttonLg: Dp = 48.dp,

    // Input heights
    val inputSm: Dp = 32.dp,
    val inputMd: Dp = 40.dp,
    val inputLg: Dp = 56.dp,

    // Avatar sizes
    val avatarSm: Dp = 32.dp,
    val avatarMd: Dp = 40.dp,
    val avatarLg: Dp = 56.dp,
    val avatarXl: Dp = 80.dp,
)

// ============================================================================
// BREAKPOINT TOKENS (for responsive design)
// ============================================================================

/**
 * Screen breakpoints for adaptive layouts
 */
@Immutable
data class BreakpointTokens(
    val mobile: Dp = 0.dp,        // 0 - 599dp
    val tablet: Dp = 600.dp,      // 600 - 839dp
    val desktop: Dp = 840.dp,     // 840 - 1199dp
    val largeDesktop: Dp = 1200.dp, // 1200+dp
) {
    enum class ScreenClass {
        Mobile,
        Tablet,
        Desktop,
        LargeDesktop
    }

    fun classifyWidth(width: Dp): ScreenClass = when {
        width < tablet -> ScreenClass.Mobile
        width < desktop -> ScreenClass.Tablet
        width < largeDesktop -> ScreenClass.Desktop
        else -> ScreenClass.LargeDesktop
    }
}

// ============================================================================
// MOTION/ANIMATION TOKENS
// ============================================================================

/**
 * Animation duration tokens
 */
@Immutable
data class MotionTokens(
    val durationInstant: Int = 0,
    val durationFast: Int = 150,
    val durationNormal: Int = 300,
    val durationSlow: Int = 500,
    val durationSlowest: Int = 700,
)

// ============================================================================
// COMBINED DESIGN TOKENS
// ============================================================================

/**
 * Complete design token set - all tokens in one place
 */
@Immutable
data class DesignTokens(
    val colors: ColorSchemeTokens,
    val palette: ColorPalette = ColorPalette(),
    val typography: TypographyTokens = TypographyTokens(),
    val spacing: SpacingTokens = SpacingTokens(),
    val shapes: ShapeTokens = ShapeTokens(),
    val elevation: ElevationTokens = ElevationTokens(),
    val sizing: SizingTokens = SizingTokens(),
    val breakpoints: BreakpointTokens = BreakpointTokens(),
    val motion: MotionTokens = MotionTokens(),
)

// ============================================================================
// COMPOSITION LOCAL
// ============================================================================

val LocalDesignTokens = staticCompositionLocalOf<DesignTokens> {
    error("No DesignTokens provided. Wrap your content in a ReaktorTheme.")
}

/**
 * Access design tokens from composable scope
 */
object Tokens {
    val current: DesignTokens
        @Composable get() = LocalDesignTokens.current

    val colors: ColorSchemeTokens
        @Composable get() = current.colors

    val typography: TypographyTokens
        @Composable get() = current.typography

    val spacing: SpacingTokens
        @Composable get() = current.spacing

    val shapes: ShapeTokens
        @Composable get() = current.shapes

    val elevation: ElevationTokens
        @Composable get() = current.elevation

    val sizing: SizingTokens
        @Composable get() = current.sizing

    val breakpoints: BreakpointTokens
        @Composable get() = current.breakpoints

    val motion: MotionTokens
        @Composable get() = current.motion
}
